### PR TITLE
Re-implement(Optimize) select function

### PR DESCRIPTION
**Pull request by hangxingliu**
Here are stages of this PR:

- [x] Utility functions and their unit tests
- [x] Replacing all operations on the `self.selections` in current component with new utility functions
- [x] Add intergration test for it
- [x] Add document to explain its menchanism

The visual tests for the utility functions in this PR: <https://3bdhlh.csb.app/>

Related issues:

- #435 
- #425 
- #414

Hi @ndrsn, this PR doesn't need to be merged before all the stages of it are complete. I create this PR  right now because it can let every developers knows the there some core code is changing and let developers have a peek first.

### CODE REVIEW

**Comment by hangxingliu** [comment]
Hi @ndrsn , How do you think this breaking changes:

In original `selections` property is a 2D array. The selection state is represented like this: `self.selections[viewRowIndex] = [0,4,5]`;

But in this PR, `selection` property is a 1D array. Its each item is a selection descriptor. like this: `self.selections[0] = {type: SelectionType.Rows, row0: 10, rows1: 20}`.

There is code to export this property to public (https://github.com/TonyGermaneri/canvas-datagrid/blob/master/lib/intf.js#L1896-L1900). 
And there are many places exported the this propert in the event to the user. like this https://github.com/TonyGermaneri/canvas-datagrid/blob/master/lib/publicMethods.js#L434-L439

Because I think it is necessary to change the implementation of `selection` property if we need to resolve issues I mentioned in this PR. This PR can't be merge if you don't accept this breaking change.

Waiting for your reply to continue changing code. Thanks

**Comment by hangxingliu** [comment]
By the way, I think the property named `columnIndex` in here https://github.com/hangxingliu/canvas-datagrid/blob/master/lib/draw.js#L1157 needs to be changed. But it is a breaking change too because the visible cells can be access by public via `grid.visibleCells`

Because in the other place of this component. `columnIndex` is used for indicating the item index in the schema array. So it can mislead the developer to use it in the wrong place.

**Comment by ndrsn** [comment]
> [](https://github.com/hangxingliu)Because in the other place of this component. columnIndex is used for indicating the item index in the schema array. So it can mislead the developer to use it in the wrong place.

Not sure I completely follow — how would you propose we rename it? And would be renaming the `columnIndex` you're pointing at here, or the other location?

**Comment by hangxingliu** [comment]
> Do you think we would be able to have both versions (two names, grid.selections and grid.<something-else> co-existing at the same time

Yep, I think it is a good idea. (Do you have any advise about naming the new property). I can define a getter for the public access. This getter can return an array like what original selection looks like from the new selection and it will print a warning on the console if anyone access it.

**Comment by hangxingliu** [comment]
> Not sure I completely follow — how would you propose we rename it? And would be renaming the columnIndex you're pointing at here, or the other location?

Because for the compatibility, I don't recommend to change it now. Because in the source code. The name `columnIndex` is used for indicating the column index of the view. (Like the variable name `viewColumnIndex `) But somtimes the name `columnIndex` is used for indicating the item index of the schema. (Like the variable name `boundColumnIndex `). I think we can discuss about this part in the future.

**Comment by ndrsn** [comment]
> Do you have any advise about naming the new property

How about `selectionList` ? Since it's temporary until we've deprecated the old interface (and will keep it a private property?), it's not hugely important to get the naming exactly right, I think.

**Comment by ndrsn** [comment]
> > Not sure I completely follow — how would you propose we rename it? And would be renaming the columnIndex you're pointing at here, or the other location?
> 
> Because for the compatibility, I don't recommend to change it now. Because in the source code. The name `columnIndex` is used for indicating the column index of the view. (Like the variable name `viewColumnIndex `) But somtimes the name `columnIndex` is used for indicating the item index of the schema. (Like the variable name `boundColumnIndex `). I think we can discuss about this part in the future.

I _think_ I understand! Ah yes, let's tackle that later — but good observation.

**Comment by ndrsn** [comment]
I do have a question regarding the `RangeSelector` type: why, if any reason, did you opt for `col0/col1/row0/row1` ? It's a bit more verbose, but would `startColumnIndex/endColumnIndex/startRowIndex/endRowIndex` (or sans -index) work in this context as well? My worry is that a name like `col0` or `col1` would lead the user to think "what about `col2` and `col3`?" — whereas start/end imply (in my mind) more that it's a bounded range. Feelings?

**Comment by hangxingliu** [comment]
> It's a bit more verbose, but would `startColumnIndex/endColumnIndex/startRowIndex/endRowIndex` (or sans -index) work in this context as well?

OK, I will replace them to these style after the features be tested. Because I was worry about the naming of properties are too long. I think we can use the naming you mentioned without `-index`.

- `startColumn`
- `endColumn`
- `startRow`
- `endRow`

**Comment by ndrsn** [comment]
> OK, I will replace them to these style after the features be tested. Because I was worry about the naming of properties are too long. I think we can use the naming you mentioned without -index.

Your suggestion sounds good!

Note: A little context regarding this: my thinking regarding has changed in this respect over the years: I used to go for terse/short names, but I now believe code is read more than it is written thus should be optimized for reading and debugging over writing, so (IMHO) the more explicit and descriptivee the variable names the better. It's not a hard rule or anything, but compared to some of the legacy variable names `c` or `s` (from the days before we had bundlers and optimizers), I believe it's an improvement.

**Comment by hangxingliu** [comment]
> I used to go for terse/short names, but I now believe code is read more than it is written thus should be optimized for reading and debugging over writing

haha, same. I remeber I wrote things like a compiled JavaScript code. But after I touched some large projects. I changed my mind like what you said. And I have also started adding more comment on things that are hard to describe by the name. Thanks.

**Comment by hangxingliu** [comment]
Hi @ndrsn , This PR is ready for review and merge. And I will write a document or blog on the next week about this new selection mechanism.

File: lib/events/index.js
Comment (ndrsn) [inline lib/events/index.js:1752]: Stray `console.log` ?

File: lib/intf.js
Comment (ndrsn) [inline lib/intf.js:1696]: Should we warn the user here, of accessing a soon-to-be-deprecated version of `selections` ?

File: lib/selections/index.js
Comment (ndrsn) [inline lib/selections/index.js:162]: @velitasali , this seems to directly impact your work on #502 — depending on the merge order of both of these PRs, one or both of you may encounter a merge conflict or lost work; perhaps you could both could take a look to see what the impact is?

File: lib/selections/index.js
Comment (ndrsn) [inline lib/selections/index.js:272]: Minor nit-pick: there is a typo in `supressEvent`; it should be `suppressEvent` (two p-s)

File: lib/selections/index.js
Comment (ndrsn) [inline lib/selections/index.js:281]: Same

File: lib/selections/index.js
Comment (ndrsn) [inline lib/selections/index.js:427]: Stray `console.log` ?

**Review by ndrsn — CHANGES_REQUESTED** [review]
@hangxingliu , some minor comments. It's a big PR (!) so I'm not sure I managed to covered all of it properly — if not, apologies. My main question here is: apart from our earlier discussion regarding `selections` and `selectionList`, is there any change in the external interface?

The other point is that yours and @velitasali's work is overlapping a bit, we might need to sort that out first before merging one or the other PR.

File: lib/selections/index.js
Comment (velitasali) [reply lib/selections/index.js:162]: @ndrsn, `fitSelectedColumns` is in `selections/index.js` now. I can rebase my commits if this is merged first. It is up to you :)

Comment (ndrsn) [reply lib/selections/index.js:162]: Ah yeah that sounds like it may be easiest, thanks! Lets wait for @hangxingliu's feedback on my review, and then see if we can merge; it'll probably be tomorrow (Tuesday).

**Comment by hangxingliu** [comment]
Hi @ndrsn, Thanks for your review and suggestions. I will modify the code tonight and ping you again

File: lib/selections/index.js
Comment (hangxingliu) [reply lib/selections/index.js:162]: @ndrsn @velitasali , I think it is no worry about the conflict on this function. I believe it can be resolved very easy. I have resolved the conflict between this PR and Veli's last PR. And I will modifiy the code in this PR tonight.

File: lib/selections/index.js
Comment (hangxingliu) [reply lib/selections/index.js:558]: lol, quite lot this typo. it is caused by "copy and paste"

File: lib/intf.js
Comment (ndrsn) [inline lib/intf.js:1701]: Great solution!

But:

`selectionsList -> selectionList`

File: lib/selections/util.js
Comment (ndrsn) [inline lib/selections/util.js:23]: Missing description?

**Review by ndrsn — CHANGES_REQUESTED** [review]
Very minor comments! Looks great otherwise!

**Comment by hangxingliu** [comment]
@ndrsn  Oh, thanks very much. missed these places.

**Comment by ndrsn** [comment]
All good to go! Thanks a lot, @hangxingliu!

**Pull request merged**
